### DISCUSSION TITLE
Upsert on arbitrary where clause without on_conflict

### DISCUSSION THREAD

**Discussion by LonelyVikingMichael** [sqlalchemy/sqlalchemy#8828] [Usage Questions]
Sometimes it is desirable to upsert without knowledge of the primary key value. I'd also like to avoid unnecessary `SELECT` statements to increase performance.
Consider the following:

```python
from sqlalchemy.dialects.postgresql import UUID, ENUM

class Category(Base):
    __tablename__ = 'category'

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4, unique=True, nullable=False)
    name = Column(String(100))


class CategoryPosition(Base):
    __tablename__ = 'category_position'

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4, unique=True, nullable=False)
    context = Column(ENUM('foo', 'bar', 'baz', name='position_context'))
    category_id = Column((UUID(as_uuid=True), ForeignKey('category.id')
    position = Column(Integer())
```
New `Category` records might be added before we get to update their position, hence the need for an upsert. A `Category` might be referenced in the `CategoryPosition` table multiple times depending on the `context`, so we can't rely on unique constraints.

I'm doing something like the following and I'm not sure if it's robust enough:
```python
from sqlalchemy import update

# try update
result = session.execute(
    update(CategoryPosition)
    .where(CategoryPosition.category_id=='...', CategoryPosition.context=='foo')
    .values(position=5)
)
if result.rowcount == 0:  # update did not succeed
    result = session.add(CategoryPosition(**args)  # create a new record
    session.commit()
```
This seems to be working, but is it reliable?
I've noticed that successful updates will return a rowcount of 1.
Is there something else I can check besides `rowcount`?

**Comment by CaselIT** [comment]
Hi,

Depending on the database in use you may be able to use insert on conflict do update
- postgresql https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#sqlalchemy.dialects.postgresql.Insert.on_conflict_do_update
- sqlite https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#sqlalchemy.dialects.sqlite.Insert.on_conflict_do_update

**Reply by zzzeek** [reply]
they need to have a unique constraint on those columns though

**Comment by zzzeek** [comment]
> A Category might be referenced in the CategoryPosition table multiple times depending on the context, so we can't rely on unique constraints.

place the unique constraint on both "context" and "category_id".  that's the candidate key for the operation you're doing.   PG's upsert should be able to target that.
